# Before you start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [1]:
import pandas as pd
import numpy as np
import os
from IPython.display import display

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the San Francisco. We will start by loading the dataset and examining its contents. 

In [2]:
data = pd.read_csv('/Users/florgarciaortiz/IRONHACK/Week_4/Day_3/Morning/lab-Hypothesis-Testing/data/Salaries.csv', low_memory = False)


Examine the `salaries` dataset using the `head` function below.

In [3]:
data.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay OvertimePay   OtherPay Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.0  400184.25      NaN  567595.43         567595.43   
1  155966.02   245131.88  137811.38      NaN  538909.28         538909.28   
2  212739.13   106088.18    16452.6      NaN  335279.91         335279.91   
3    77916.0    56120.71   198306.9      NaN  332343.61         332343.61   
4   134401.6      9737.0  182234.59      NaN  326373.19         326373.19   

   Year  Notes         Agency Status  
0  2011    NaN  San Francisco    NaN  
1  2011    NaN  San Francisco    NaN  
2  2011    NaN  San Francisco    NaN  
3  2011    NaN  San Francisco    NaN  
4  2011    NaN  San Francisco    NaN

We see from looking at the `head` function that there is quite a bit of missing data. Get the amount of missing data in every column

In [4]:
data.isna().sum()

Id                       0
EmployeeName             0
JobTitle                 0
BasePay                605
OvertimePay              0
OtherPay                 0
Benefits             36159
TotalPay                 0
TotalPayBenefits         0
Year                     0
Notes               148654
Agency                   0
Status              110535
dtype: int64

Get the shape of the dataframe

In [5]:
display(data.shape)

(148654, 13)

Given output of the previous two cells, drop the corresponding column and compute again the amount of missing values.

In [6]:
data.drop('Notes', axis=1, inplace=True)
data.isna().sum()

Id                       0
EmployeeName             0
JobTitle                 0
BasePay                605
OvertimePay              0
OtherPay                 0
Benefits             36159
TotalPay                 0
TotalPayBenefits         0
Year                     0
Agency                   0
Status              110535
dtype: int64

Check out what are the possible values of the column "Status".

In [7]:
data['Status'].unique()

array([nan, 'PT', 'FT'], dtype=object)

Drop any row with missing values in the "Status" column and compute again the number of missing values.

In [8]:
data.dropna(subset=['Status'], inplace=True)

In [9]:
data.isna().sum()

Id                  0
EmployeeName        0
JobTitle            0
BasePay             0
OvertimePay         0
OtherPay            0
Benefits            0
TotalPay            0
TotalPayBenefits    0
Year                0
Agency              0
Status              0
dtype: int64

Check out the types of each column and see if they make sense.

In [10]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 38119 entries, 110531 to 148653
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Id                38119 non-null  int64  
 1   EmployeeName      38119 non-null  object 
 2   JobTitle          38119 non-null  object 
 3   BasePay           38119 non-null  object 
 4   OvertimePay       38119 non-null  object 
 5   OtherPay          38119 non-null  object 
 6   Benefits          38119 non-null  object 
 7   TotalPay          38119 non-null  float64
 8   TotalPayBenefits  38119 non-null  float64
 9   Year              38119 non-null  int64  
 10  Agency            38119 non-null  object 
 11  Status            38119 non-null  object 
dtypes: float64(2), int64(2), object(8)
memory usage: 3.8+ MB


In [11]:
data.head()

Id          EmployeeName                  JobTitle    BasePay  \
110531  110532           David Shinn            Deputy Chief 3  129150.01   
110532  110533            Amy P Hart         Asst Med Examiner  318835.49   
110533  110534  William J Coaker Jr.  Chief Investment Officer  257340.00   
110534  110535        Gregory P Suhr           Chief of Police  307450.04   
110535  110536  Joanne M Hayes-White    Chief, Fire Department  302068.00   

       OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
110531        0.00  342802.63  38780.04  471952.64         510732.68  2014   
110532    10712.95   60563.54  89540.23  390111.98         479652.21  2014   
110533        0.00   82313.70  96570.66  339653.70         436224.36  2014   
110534        0.00   19266.72  91302.46  326716.76         418019.22  2014   
110535        0.00   24165.44  91201.66  326233.44         417435.10  2014   

               Agency Status  
110531  San Francisco     PT  
110532  San Francisco     FT  
110533  San Francisco     PT  
110534  San Francisco     FT  
110535  San Francisco     FT

BasePay, OvertimePay, OtherPay and Benefits must be numerical variables.

Do any type conversions and reset the index.

In [12]:
data = data.astype({'BasePay': 'float64', 'OvertimePay': 'float64', 'OtherPay': 'float64', 'Benefits': 'float64'})
data = data.reset_index(drop=True)

In [13]:
data

Id          EmployeeName                     JobTitle    BasePay  \
0      110532           David Shinn               Deputy Chief 3  129150.01   
1      110533            Amy P Hart            Asst Med Examiner  318835.49   
2      110534  William J Coaker Jr.     Chief Investment Officer  257340.00   
3      110535        Gregory P Suhr              Chief of Police  307450.04   
4      110536  Joanne M Hayes-White       Chief, Fire Department  302068.00   
...       ...                   ...                          ...        ...   
38114  148646      Carolyn A Wilson    Human Services Technician       0.00   
38115  148648        Joann Anderson  Communications Dispatcher 2       0.00   
38116  148649           Leon Walker                    Custodian       0.00   
38117  148650         Roy I Tillery                    Custodian       0.00   
38118  148654             Joe Lopez   Counselor, Log Cabin Ranch       0.00   

       OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0             0.00  342802.63  38780.04  471952.64         510732.68  2014   
1         10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2             0.00   82313.70  96570.66  339653.70         436224.36  2014   
3             0.00   19266.72  91302.46  326716.76         418019.22  2014   
4             0.00   24165.44  91201.66  326233.44         417435.10  2014   
...            ...        ...       ...        ...               ...   ...   
38114         0.00       0.00      0.00       0.00              0.00  2014   
38115         0.00       0.00      0.00       0.00              0.00  2014   
38116         0.00       0.00      0.00       0.00              0.00  2014   
38117         0.00       0.00      0.00       0.00              0.00  2014   
38118         0.00    -618.13      0.00    -618.13           -618.13  2014   

              Agency Status  
0      San Francisco     PT  
1      San Francisco     FT  
2      San Francisco     PT  
3      San Francisco     FT  
4      San Francisco     FT  
...              ...    ...  
38114  San Francisco     PT  
38115  San Francisco     PT  
38116  San Francisco     PT  
38117  San Francisco     PT  
38118  San Francisco     PT  

[38119 rows x 12 columns]

Check out if "TotalPayBenefits" = "BasePay" + "OvertimePay" + "OtherPay" + "Benefits"

In [14]:
is_equal = (data['BasePay'] + data['OvertimePay'] + data['OtherPay'] + data['Benefits']) == data['TotalPayBenefits']
num_not_equal = len(is_equal[is_equal == False])
num_not_equal

10022

In [15]:
print(f"Number of rows where the sum is not equal to 'TotalPayBenefits': {num_not_equal}")

Number of rows where the sum is not equal to 'TotalPayBenefits': 10022


What is the percetage of employees for which the previous assumption is not True?

In [16]:
total_count = len(data['TotalPayBenefits'])
ratio = (num_not_equal / total_count) * 100
ratio

26.29135076995724

There are different departments in the city. List all departments and the count of employees in each department.

In [17]:
data['JobTitle'].unique()

array(['Deputy Chief 3', 'Asst Med Examiner', 'Chief Investment Officer',
       'Chief of Police', 'Chief, Fire Department', 'Dept Head V',
       'Executive Contract Employee', 'Battalion Chief, Fire Suppress',
       'Asst Chf of Dept (Fire Dept)', 'Gen Mgr, Public Trnsp Dept',
       'Senior Physician Specialist', 'Commander 3',
       'Dep Chf of Dept (Fire Dept)', 'Administrator, DPH', 'Mayor',
       'Adm, SFGH Medical Center', 'Controller', 'Anesthetist',
       'Dep Dir V', 'Port Director', 'Manager, Dept Public Health',
       'Deputy Sheriff', 'Assistant Deputy Chief 2', 'Deputy Dir II, MTA',
       'Firefighter', 'Emergency Medical Svcs Chief', 'Manager VIII',
       'District Attorney', 'Inspector 3', 'Captain 3',
       'Captain, Fire Suppression', 'Nursing Supervisor',
       'Lieut,Fire Prev', 'Dept Head IV', 'City Attorney',
       'EMT/Paramedic/Firefighter', 'Police Officer 3',
       'Lieutenant, Fire Suppression', 'Sheriff (SFERS)', 'Dep Dir IV',
       'Nurse Mana

In [18]:
department_counts = data.groupby('JobTitle')['EmployeeName'].count()
print(department_counts)

JobTitle
ACPO,JuvP, Juv Prob (SFERS)        1
ASR Senior Office Specialist      22
ASR-Office Assistant              15
Account Clerk                     93
Accountant I                       2
                                  ..
Wire Rope Cable Maint Sprv         1
Worker's Comp Supervisor 1         6
Worker's Compensation Adjuster    26
X-Ray Laboratory Aide             35
Youth Comm Advisor                 1
Name: EmployeeName, Length: 996, dtype: int64


# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of **all FT workers is significantly different from $75/hr**. Get first the hourly wage by dividing "TotalPayBenefits" by 50 weeks (assuming 10 labour days of holidays) and by 40hrs (assuming a 40hrs week).

$$Hourly Wage = \frac{TotalPayBenefits}{1 year}\frac{1 year}{50 Week}\frac{1 Week}{40 hr}$$

Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [19]:
data1 = data[data['Status'] == 'FT'].assign(HourlyWage=data['TotalPayBenefits'] / 2000)

In [20]:
data1

Id          EmployeeName                      JobTitle    BasePay  \
1      110533            Amy P Hart             Asst Med Examiner  318835.49   
3      110535        Gregory P Suhr               Chief of Police  307450.04   
4      110536  Joanne M Hayes-White        Chief, Fire Department  302068.00   
5      110537       Ellen G Moffatt             Asst Med Examiner  270222.04   
6      110538         John L Martin                   Dept Head V  311298.55   
...       ...                   ...                           ...        ...   
28535  139067         Nasan M Snell  Public Svc Aide-Public Works   26653.98   
28824  139356    Eddie Ellenberg Iv  Environmental Service Worker   31856.71   
28873  139405       Bernard O Sices  Environmental Service Worker   31592.80   
29324  139856          Julie Y Zhen  Public Svc Aide-Public Works   27980.52   
29794  140326         Camea L Davis  Public Svc Aide-Public Works   27116.62   

       OvertimePay  OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
1         10712.95  60563.54  89540.23  390111.98         479652.21  2014   
3             0.00  19266.72  91302.46  326716.76         418019.22  2014   
4             0.00  24165.44  91201.66  326233.44         417435.10  2014   
5          6009.22  67956.20  71580.48  344187.46         415767.94  2014   
6             0.00      0.00  89772.32  311298.55         401070.87  2014   
...            ...       ...       ...        ...               ...   ...   
28535         0.00      0.00  18032.01   26653.98          44685.99  2014   
28824      1529.52    350.99   7916.24   33737.22          41653.46  2014   
28873      1274.80    178.50   8144.01   33046.10          41190.11  2014   
29324       349.54    777.30   7568.15   29107.36          36675.51  2014   
29794         0.00      0.00   4856.19   27116.62          31972.81  2014   

              Agency Status  HourlyWage  
1      San Francisco     FT  239.826105  
3      San Francisco     FT  209.009610  
4      San Francisco     FT  208.717550  
5      San Francisco     FT  207.883970  
6      San Francisco     FT  200.535435  
...              ...    ...         ...  
28535  San Francisco     FT   22.342995  
28824  San Francisco     FT   20.826730  
28873  San Francisco     FT   20.595055  
29324  San Francisco     FT   18.337755  
29794  San Francisco     FT   15.986405  

[22334 rows x 13 columns]

In [21]:
# Your code here: (Compute the mean hourly wage for all the "FT" employees)
HourlyWage = data1['TotalPayBenefits'] / 2000
display(data1[data1['Status']== 'FT']['HourlyWage'].mean())

69.264206407271

In [22]:
# Your code here: (compute the t_statistic). Take into account that this dataset is a sample of a real population.
# Remember that you only need to consider "FT" employees

#H0: HourlyWage in FT = $75/hr
#H1: HourlyWate in FT != $75/hr

# Method 1: Critical value. Get the critical value and compare it against your statisttic.
# Your code here: 

from scipy.stats import t

ft_data = data1[data1['Status'] == 'FT']

population_mean = data1['HourlyWage'].mean()
sample_size = 30
sample_mean = data1['HourlyWage'].sample(30, random_state=1).mean()
sample_std = data1['HourlyWage'].sample(30, random_state=1).std()
standard_error = sample_std / (sample_size ** 0.5)

degrees_of_freedom = sample_size - 1

t_statistic = (sample_mean - population_mean) / standard_error


alpha = 0.05
critical_value = t.ppf(1 - alpha / 2, degrees_of_freedom)

print("population_mean:", population_mean)
print("sample_mean:", sample_mean)
print("sample_std:", sample_std)
print("standard_error:", standard_error)
print("t-statistic:", t_statistic)
print("critical_value:", critical_value)

population_mean: 69.264206407271
sample_mean: 71.23279016666667
sample_std: 25.910411696774474
standard_error: 4.730572320189696
t-statistic: 0.41614071747596293
critical_value: 2.045229642132703


In [23]:
# Method 2: Use the p-value method.
# Your code here:

from scipy.stats import t

t_statistic = 0.41614071747596293
degrees_of_freedom = 29  

p_value = 2 * (1 - t.cdf(abs(t_statistic), degrees_of_freedom))

print("p-value:", p_value)

p-value: 0.6803707335778229


In [24]:
# Method 3: Use the ttest_1samp function from scipy. 
# Check the documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html)
# Make sure that you have a scipy version >=1.6.0. If that's not your case please ugrade your scipy version using
# !pip install -U scipy
# Your code here:

import scipy.stats as st

HourlyWage = data1[data1['Status']== 'FT']['HourlyWage'].sample(30, random_state=1)
st.ttest_1samp(HourlyWage,75)

Ttest_1sampResult(statistic=-0.7963539247154477, pvalue=0.4322975682437393)

Are all the methods in agreement?

With the 3 methods we accept the null hypothesis. So that the hourly wage of all FT workers is significantly different from $75/hr

We are also curious about salaries in the police force. The chief of police in Chicago claimed in a press briefing that salaries this year are **higher than last year's mean of $86000/year for all salaried employees** (use the column "TotalPayBenefits". Test  hypothesis using a 95% confidence interval.

Hint: Use apply and a lambda function to check in "Police" is in the "JobTitle" to get all the "Police" jobs.

In [25]:
# Your code here: (compute the t_statistic). Take into account that this dataset is a sample of a real population.
# Remember that you only need to consider "Police" employees

#H0: Salaries in police force = $86000
#H1: Salaries in police force > $86000

# Method 1: Critical value. Get the critical value and compare it against your statisttic.
# Your code here: 

from scipy.stats import t

police_jobs = data[data['JobTitle'].apply(lambda x: 'Police' in x)]

population_mean = police_jobs['TotalPayBenefits'].mean()
sample_size = 30
sample_mean = police_jobs['TotalPayBenefits'].sample(30, random_state=1).mean()
sample_std = police_jobs['TotalPayBenefits'].sample(30, random_state=1).std()
standard_error = sample_std / (sample_size ** 0.5)

degrees_of_freedom = sample_size - 1

t_statistic = (sample_mean - population_mean) / standard_error

alpha = 0.05
critical_value = t.ppf(1 - alpha, degrees_of_freedom)

print("population_mean:", population_mean)
print("sample_mean:", sample_mean)
print("sample_std:", sample_std)
print("standard_error:", standard_error)
print("t-statistic:", t_statistic)
print("critical_value:", critical_value)


population_mean: 148349.04212519721
sample_mean: 158041.32166666666
sample_std: 39448.05509087852
standard_error: 7202.196540993557
t-statistic: 1.3457393846867127
critical_value: 1.6991270265334972


In [26]:
# Method 2: Use the p-value method.
# Your code here:

In [27]:
from scipy.stats import t

t_statistic = 1.3457393846867127
degrees_of_freedom = 29  

p_value = 1 - t.cdf(t_statistic, degrees_of_freedom)

print("p-value:", p_value)

p-value: 0.09440709956725468


In [28]:
# Method 3: Use the ttest_1samp function from scipy. 
# Check the documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html)
# Make sure that you have a scipy version >=1.6.0. If that's not your case please ugrade your scipy version using
# !pip install -U scipy
# Your code here:
import scipy.stats as stats

police_jobs = data[data['JobTitle'].apply(lambda x: 'Police' in x)]
mean_police_salary = police_jobs['TotalPayBenefits'].sample(30, random_state=1)
print('p value (single tailed): ',st.ttest_1samp(mean_police_salary, 86000, alternative="greater")[1]/2)

p value (single tailed):  1.6399365619098177e-11


With the methods 1 and 2 we accept the null hypothesis. But, with the method 3 we reject the null hypothesis, and we accept that salaries in police force are higher than 86,000.

The workers from the "JobTitle" with the most employees have complained that their hourly wage is **less than $35/hour**. Using a one sample t-test, test this one-sided hypothesis at the 95% confidence level.

In [29]:
# Your code here: (Get the department which has most employees)

In [30]:
department_counts = data.groupby('JobTitle')['EmployeeName'].count().sort_values(ascending=False)
print(department_counts)

JobTitle
Transit Operator                  2479
Special Nurse                     1478
Registered Nurse                  1234
Public Svc Aide-Public Works       916
Firefighter                        815
                                  ... 
Marina Associate Manager             1
Marina Assistant Manager             1
Light Rail Vehicle Equip Eng         1
Lieutenant (Police Department)       1
Youth Comm Advisor                   1
Name: EmployeeName, Length: 996, dtype: int64


In [31]:
# Your code here: (compute the t_statistic). Take into account that this dataset is a sample of a real population.
# Remember that you only need to consider the right "JobTitle" employees

In [32]:
# The most frequent title is Transit Operator.
#H0: Salaries for transit operator = $35/hour
#H1: Salaries for transit operator < $35/hour

# Method 1: Critical value. Get the critical value and compare it against your statisttic.
# Your code here: 

from scipy.stats import t

data2 = data.assign(HourlyWage=data['TotalPayBenefits'] / 2000)
transit_operator = data2[data2['JobTitle'].apply(lambda x: 'Transit Operator' in x)]
population_mean = transit_operator['HourlyWage'].mean()
sample_size = 30
sample_mean = transit_operator['HourlyWage'].sample(30, random_state=1).mean()
sample_std = transit_operator['HourlyWage'].sample(30, random_state=1).std()
standard_error = sample_std / (sample_size ** 0.5)

degrees_of_freedom = sample_size - 1

t_statistic = (sample_mean - population_mean) / standard_error

alpha = 0.05
critical_value = t.ppf(1 - alpha, degrees_of_freedom)

print("population_mean:", population_mean)
print("sample_mean:", sample_mean)
print("sample_std:", sample_std)
print("standard_error:", standard_error)
print("t-statistic:", t_statistic)
print("critical_value:", critical_value)

population_mean: 43.35949773900775
sample_mean: 45.87508383333332
sample_std: 19.68028627288567
standard_error: 3.5931122432729174
t-statistic: 0.700113418119707
critical_value: 1.6991270265334972


In [33]:
# Method 2: Use the p-value method.
# Your code here:
from scipy.stats import t

t_statistic = 0.700113418119707
degrees_of_freedom = 29  

p_value = 1 - t.cdf(t_statistic, degrees_of_freedom)

print("p-value:", p_value)

p-value: 0.24471768597791155


In [34]:
# Method 3: Use the ttest_1samp function from scipy. 
# Check the documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html)
# Make sure that you have a scipy version >=1.6.0. If that's not your case please ugrade your scipy version using
# !pip install -U scipy
# Your code here:

In [35]:
import scipy.stats as stats

transit_operator = data2[data2['JobTitle'].apply(lambda x: 'Transit Operator' in x)]
mean_transit_operator = transit_operator['HourlyWage'].sample(30, random_state=1)
print('p value (single tailed): ',st.ttest_1samp(mean_police_salary, 86000, alternative="less")[1]/2)

p value (single tailed):  0.4999999999836006


With the 3 methods we accept the null hypothesis. So we accept that salaries for transit operators are equal to $35/hour.

# Challenge 3: To practice - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level.

In [36]:
# Your code here:
from scipy.stats import t

hourly_wage = data1['HourlyWage']

sample_mean = hourly_wage.mean()
sample_std = hourly_wage.std()
sample_size = 30

standard_error = sample_std / (sample_size ** 0.5)


degrees_of_freedom = sample_size - 1

# Calculate critical value for a 95% confidence interval
alpha = 0.05
critical_value = t.ppf(1 - alpha / 2, degrees_of_freedom)

margin_of_error = critical_value * standard_error

# Calculate lower and upper bounds of the confidence interval
lower_bound = sample_mean - margin_of_error
upper_bound = sample_mean + margin_of_error


print("95% Confidence Interval for Mean Hourly Wage:")
print("Lower Bound:", lower_bound)
print("Upper Bound:", upper_bound)
print("sample mean:", sample_mean)


95% Confidence Interval for Mean Hourly Wage:
Lower Bound: 60.32500875462077
Upper Bound: 78.20340405992124
sample mean: 69.264206407271


Now compute a 95% confidence interval for the hourly salary of all the Police employees.

In [37]:
# Your code here:
from scipy.stats import t

police_jobs = data2[data2['JobTitle'].apply(lambda x: 'Police' in x)]
salary_police = police_jobs['HourlyWage']

sample_mean = salary_police.mean()
sample_std = salary_police.std()
sample_size = 30

standard_error = sample_std / (sample_size ** 0.5)


degrees_of_freedom = sample_size - 1

# Calculate critical value for a 95% confidence interval
alpha = 0.05
critical_value = t.ppf(1 - alpha / 2, degrees_of_freedom)

margin_of_error = critical_value * standard_error

# Calculate lower and upper bounds of the confidence interval
lower_bound = sample_mean - margin_of_error
upper_bound = sample_mean + margin_of_error


print("95% Confidence Interval for Mean Hourly Wage:")
print("Lower Bound:", lower_bound)
print("Upper Bound:", upper_bound)
print("sample mean:", sample_mean)


95% Confidence Interval for Mean Hourly Wage:
Lower Bound: 64.07477537992378
Upper Bound: 84.27426674527322
sample mean: 74.1745210625985


# Chi2 test

Now we want to know if the amount of full time "FT" and part time "PT" employees is equal between Lawers, Meds, Police, Firemen and other departments. 

Considering all the options in this groups of employees will be very time consuming. To simplify this process, create first a function that returns:

* "Policemen" if "Police" is found on "JobTitle"
* "Firemen" if "Fire" is found on "JobTitle"
* "Medical" if "Med" or "Nurse" is found on "JobTitle"
* "Lawyer" if "Attorney" is found on "JobTitle"
* "Other" in any other cases

Then, create a new column named "employee_group" that determines to which group belong the employee. 

In [38]:
# Your code here: (Create the function)
def categories(title):
    if "Police" in title:
        return "Policemen"
    elif "Fire" in title:
        return "Firemen"
    elif "Med" in title or "Nurse" in title:
        return "Medical"
    elif "Attorney" in title:
        return "Lawyer"
    else:
        return "Other"

data['employee_group'] = data['JobTitle'].apply(lambda x: categories(x))

In [39]:
data

Id          EmployeeName                     JobTitle    BasePay  \
0      110532           David Shinn               Deputy Chief 3  129150.01   
1      110533            Amy P Hart            Asst Med Examiner  318835.49   
2      110534  William J Coaker Jr.     Chief Investment Officer  257340.00   
3      110535        Gregory P Suhr              Chief of Police  307450.04   
4      110536  Joanne M Hayes-White       Chief, Fire Department  302068.00   
...       ...                   ...                          ...        ...   
38114  148646      Carolyn A Wilson    Human Services Technician       0.00   
38115  148648        Joann Anderson  Communications Dispatcher 2       0.00   
38116  148649           Leon Walker                    Custodian       0.00   
38117  148650         Roy I Tillery                    Custodian       0.00   
38118  148654             Joe Lopez   Counselor, Log Cabin Ranch       0.00   

       OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  Year  \
0             0.00  342802.63  38780.04  471952.64         510732.68  2014   
1         10712.95   60563.54  89540.23  390111.98         479652.21  2014   
2             0.00   82313.70  96570.66  339653.70         436224.36  2014   
3             0.00   19266.72  91302.46  326716.76         418019.22  2014   
4             0.00   24165.44  91201.66  326233.44         417435.10  2014   
...            ...        ...       ...        ...               ...   ...   
38114         0.00       0.00      0.00       0.00              0.00  2014   
38115         0.00       0.00      0.00       0.00              0.00  2014   
38116         0.00       0.00      0.00       0.00              0.00  2014   
38117         0.00       0.00      0.00       0.00              0.00  2014   
38118         0.00    -618.13      0.00    -618.13           -618.13  2014   

              Agency Status employee_group  
0      San Francisco     PT          Other  
1      San Francisco     FT        Medical  
2      San Francisco     PT          Other  
3      San Francisco     FT      Policemen  
4      San Francisco     FT        Firemen  
...              ...    ...            ...  
38114  San Francisco     PT          Other  
38115  San Francisco     PT          Other  
38116  San Francisco     PT          Other  
38117  San Francisco     PT          Other  
38118  San Francisco     PT          Other  

[38119 rows x 13 columns]

Determine how many "PT" and "FT" employess have all the employees groups.

In [40]:
# Your code here: (Store the output dataframe into a new variable)
employee_counts = data.groupby(['employee_group', 'Status']).size().unstack()

employee_per_status = employee_counts.copy()

employee_per_status

Status             FT     PT
employee_group              
Firemen          1333    178
Lawyer            317    102
Medical          1028   2889
Other           18126  12245
Policemen        1530    371

Now try compute the expected frequencies doing the calculations with the individual probabilities. Remember that the Chi2 test assumes that both variables (employee_group and FT/PT) are not related (therefore they are independent). Therefore, to compute the expected frequencies you need to compute the probability of each cell and multiply it by the number of observations. ie:

$$\nu(x,y) = p(x,y) * N = p(x) * p(y) * N$$

bear in mind that in general: $p(x,y)\neq p(x)*p(y)$; the equality will only be true if x and y are independent. However, the null hypotheses says that **x and y are independent.** but that's what we're assuming with the null hypotheses.

where "x" is the "employee_group" and "y" the (FT/PT). 

In [41]:
# Create an empty dataframe named "frequencies" to store the data.
# Your code here:

probabilities = pd.DataFrame({
    'FT': employee_per_status['FT'] / (sum(employee_per_status['FT']) + sum(employee_per_status['PT'])),
    'PT': employee_per_status['PT'] / (sum(employee_per_status['FT']) + sum(employee_per_status['PT']))
})
frequencies = pd.DataFrame({
    'FT_freq': employee_per_status['FT'] * probabilities['FT'],
    'PT_freq': employee_per_status['PT'] * probabilities['PT']
})

frequencies = pd.concat([probabilities, frequencies], axis=1)
display(frequencies)

# In the next points I just add the result, because the code is this.

FT        PT      FT_freq      PT_freq
employee_group                                              
Firemen         0.034969  0.004670    46.614261     0.831187
Lawyer          0.008316  0.002676     2.636192     0.272935
Medical         0.026968  0.075789    27.723288   218.954353
Other           0.475511  0.321231  8619.110575  3933.472153
Policemen       0.040137  0.009733    61.410320     3.610824

In [42]:
# Your code here: Compute Expected frequency of being "Firemen" and "FT". Store the solution in a variable named "firemen_ft"

46.614261

In [43]:
# Your code here: Compute Expected frequency of being "Firemen" and "PT". Store the solution in a variable named "firemen_pt"

0.831187

In [44]:
# Your code here: Compute Expected frequency of being "Lawyers" and "FT". Store the solution in a variable named "lawyers_ft"

2.636192

In [45]:
# Your code here: Compute Expected frequency of being "Lawyers" and "PT". Store the solution in a variable named "lawyers_pt"

0.272935

In [46]:
# Your code here: Compute Expected frequency of being "Medical" and "FT". Store the solution in a variable named "medical_ft"

27.723288

In [47]:
# Your code here: Compute Expected frequency of being "Medical" and "PT". Store the solution in a variable named "medical_pt"

218.954353

In [48]:
# Your code here: Compute Expected frequency of being "Other" and "FT". Store the solution in a variable named "other_ft"

8619.110575

In [49]:
# Your code here: Compute Expected frequency of being "Other" and "PT". Store the solution in a variable named "other_pt"

3933.472153

In [50]:
# Your code here: Compute Expected frequency of being "Policement" and "FT". Store the solution in a variable named "policemen_ft"

61.410320

In [51]:
# Your code here: Compute Expected frequency of being "Policement" and "PT". Store the solution in a variable named "policemen_pt"

3.610824

* Store all the expected frequencies of "FT" employees in a list 
* Store all the "PT" employees into another list
* Create a dictionary with "FT" and "PT" as keys and as the values use the previous lists
* Create a dataframe with this dictionary using pd.DataFrame()

In [52]:
# Your code here:

# Store expected frequencies of "FT" employees in a list
ft_expected_freq = employee_per_status['FT'] * probabilities['FT']

In [53]:
# Store "PT" employees in another list
pt_expected_freq = employee_per_status['PT'] * probabilities['PT']


In [54]:
# Create a dictionary with "FT" and "PT" as keys and the lists as values
expected_freq_dict = {'FT': ft_expected_freq.tolist(), 'PT': pt_expected_freq.tolist()}

In [55]:
# Create a dataframe with the dictionary
frequencies = pd.DataFrame(expected_freq_dict)
# Display the dataframe
display(frequencies)

FT           PT
0    46.614261     0.831187
1     2.636192     0.272935
2    27.723288   218.954353
3  8619.110575  3933.472153
4    61.410320     3.610824

Now use the "st.chi2_contingency()" from scipy.stats [documentation here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html) to conduct a Chi2 test to determine if the diferences between employee groups are statistically significant using a 95% confidence level. Hint: fill the function with a dataframe of actual frequencies.

In [56]:
# Your code here: (use the st.chi2_contingency() function from scipy.stats to compute:
# The Chi2 value
# The p-valueYea we 
# The expected frequencies.
# Ho: there is no relationship
# Ha: there is relationship differences
# p_value = P(table | Ho) = P(table | no relationship) = 1.51e-6 < 0.05

In [57]:
import scipy.stats as stats
from scipy.stats import chi2_contingency

chi2_stat, p_value, dof, expected_chi2 = stats.chi2_contingency(employee_per_status)

print("Chi-square statistic:", chi2_stat)
print("Degrees of freedom:", dof)
print("p-value:", p_value)


Chi-square statistic: 2676.642333711905
Degrees of freedom: 4
p-value: 0.0


Check if your expected frequencies aggree with the ones obtained with the st.chi2_contingency() function

In [59]:
expected_freq_df = pd.DataFrame(expected_chi2, index=employee_per_status.index, columns=['FT_freq', 'PT_freq'])

# Compare expected frequencies
print("Expected Frequencies:")
print(expected_freq_df)
print("\nExpected Frequencies from chi2_contingency():")
print(pd.DataFrame(expected_chi2, index=employee_per_status.index, columns=['FT_freq', 'PT_freq']))


Expected Frequencies:
                     FT_freq       PT_freq
employee_group                            
Firemen           885.297988    625.702012
Lawyer            245.492956    173.507044
Medical          2294.978305   1622.021695
Other           17794.430966  12576.569034
Policemen        1113.799785    787.200215

Expected Frequencies from chi2_contingency():
                     FT_freq       PT_freq
employee_group                            
Firemen           885.297988    625.702012
Lawyer            245.492956    173.507044
Medical          2294.978305   1622.021695
Other           17794.430966  12576.569034
Policemen        1113.799785    787.200215
